nu_stats/fqutils.py

```diff
@@ -0,0 +1,772 @@
+        if self.spacial_only:
+            return spacial_factor
+        else:
+            assert hasattr(self, 'energy_likelihood'), 'Missing P(E_d|γ, src)'
+            energy_factor = self.energy_likelihood(E_d, gamma) # p(E_d|gamma)
+            if energy_factor == 0: raise ValueError(
+                f'P(E_d|Υ,src) gave 0.0 for E_d={E_d}, Υ={gamma}')
+            return spacial_factor * energy_factor
+
+    def bg_likelihood(self, E_d, gamma):
+        """ Evaluate Bg likelihood
+        Args:
+            source_dir (np.ndarray): direction of source
+            obs_dir (np.ndarray): direction of observed neutrino
+            kappa (float): uncertainty of neutrino direction in obs
+        """
+        spacial_factor = 1/(4*np.pi) # uniform on sphere
+        if self.spacial_only:
+            return spacial_factor
+        elif hasattr(self, 'bg_energy_likelihood'):
+            return spacial_factor * self.bg_energy_likelihood(E_d, gamma)    
+
+    if True: # Foldable section for methods on individual events
+        def test_stat(self,
+            E_d : np.ndarray, # Edets from sim
+            obs_dir: np.ndarray, # det_dir from sim
+            gamma: float,
+            source_dir: np.ndarray,
+            kappa: float # Detector
+        ):
+            N_obs = obs_dir.shape[0]
+            TS = np.empty(N_obs)
+            TS[:] = np.NaN
+            for j in range(N_obs):
+                S = self.source_likelihood(
+                    E_d[j],
+                    obs_dir[j],
+                    gamma,
+                    source_dir,
+                    kappa
+                    )
+                LR = S / self.bg_likelihood(E_d[j], gamma)
+                TS[j] = 2*np.log(LR)
+            return TS
+
+        def event_statistics(self,
+            fit_input: OrderedDict,
+            bg_dat: OrderedDict,
+            gamma: float
+        ):
+            """Get p-values for fit_input events coming from bg
+            Args:
+                fit_input (OrderedDict): output from Simulation.get_data()
+                bg_dat (OrderedDict): output from background Simulation.get_data()
+                gamma (float): 
+
+            Returns:
+                np.ndarray: 
+            """
+            self.fit_input = fit_input
+
+            sim_TS = self.test_stat(
+                self.fit_input['Edet'],
+                self.fit_input['det_dir'],
+                gamma,
+                self.fit_input['source_dir'],
+                self.fit_input['kappa']
+                )
+            bg_TS = self.test_stat(
+                bg_dat['Edet'],
+                bg_dat['det_dir'],
+                gamma,
+                self.fit_input['source_dir'],
+                bg_dat['kappa']
+                )
+
+            ## get p values
+            sim_p = np.zeros_like(sim_TS)
+            for i,obs_TS in enumerate(sim_TS):
+                sim_p[i] = (np.mean(bg_TS >= obs_TS))
```

Comment:
Why is this the mean and not the integral? Let's talk about it in the next meeting.